pkg/odo/cli/list/list.go

```diff
@@ -0,0 +1,279 @@
+	lo.devfilePath = location.DevfileLocation("")
+	if util.CheckPathExists(lo.devfilePath) {
+
+		// Set the project / namespace based on the devfile context
+		lo.project = lo.Context.GetProject()
+
+		// Parse the devfile
+		devObj, parseErr := devfile.ParseAndValidateFromFile(lo.devfilePath)
+		if parseErr != nil {
+			return parseErr
+		}
+
+		// Convert the devfile to a Component{} for table output and manipulation
+		// we want to limit using lo.EnvSpecificInfo, but fortunately, it will simply try to find
+		// lo.devfilePath for our usage.
+		localComponent, err := component.GetComponentFromDevfile(devObj, lo.project)
+		if err != nil {
+			return err
+		}
+		lo.localComponent = localComponent
+
+	}
+
+	// If the context is "" (devfile.yaml not found..), we get the active one from KUBECONFIG.
+	if lo.project == "" {
+		lo.project = lo.clientset.KubernetesClient.GetCurrentNamespace()
+	}
+
+	// If the namespace flag has been passed, we will search there.
+	// if it hasn't, we will search from the default project / namespace.
+	if lo.namespaceFlag != "" {
+		lo.namespaceFilter = lo.namespaceFlag
+	} else {
+		lo.namespaceFilter = lo.project
+	}
+
+	return
+}
+
+// Validate ...
+func (lo *ListOptions) Validate() (err error) {
+	return nil
+}
+
+// Run has the logic to perform the required actions as part of command
+func (lo *ListOptions) Run() error {
+
+	// All variables pertaining to output / combination
+	devfileComponents := []component.Component{}
+
+	// Step 1.
+	// Retrieve all related components from the Kubernetes cluster, from the given namespace
+	if lo.clientset.KubernetesClient != nil {
+
+		// Create a spinner since this function takes up 99% of `odo list` time
+		listSpinner := log.Spinnerf("Listing components from namespace '%s'", lo.namespaceFilter)
+		defer listSpinner.End(false)
+
+		var err error
+		devfileComponents, err = component.ListAllClusterComponents(lo.clientset.KubernetesClient, lo.namespaceFilter)
+		if err != nil {
+			return err
+		}
+		listSpinner.End(true)
+	}
+
+	// Step 2.
+	// If we have a local component, let's add it to the list of Devfiles
+	// This checks lo.localComponent.Name. If it's empty, we didn't parse one in the Complete() function, so there is no local devfile.
+	// We will only append the local component to the devfile if it doesn't exist in the list.
+	if (lo.localComponent.Name != "") && !component.Contains(lo.localComponent, devfileComponents) {
+		devfileComponents = append(devfileComponents, lo.localComponent)
+	}
+
+	if log.IsJSON() {
+		machineoutput.OutputSuccess(devfileComponents)
+	} else {
+		lo.HumanReadableOutput(log.GetStdout(), devfileComponents)
+	}
+
+	return nil
+}
+
+// NewCmdList implements the list odo command
+func NewCmdList(name, fullName string) *cobra.Command {
+	o := NewListOptions()
+
+	var listCmd = &cobra.Command{
+		Use:         name,
+		Short:       "List all components in the current application",
```

Comment:
better would be "in the current namespace" "in the current application" doesn't make sense and it is not true.